### DISCUSSION TITLE
Audio only from Twitch streams

### DISCUSSION THREAD

**Discussion by Drizzt321** [streamlink/streamlink#5284] [Q&A] [answered]
### Description

I'm exciting to find this project, after thinking about trying to do something like this for a long time.

So as the title says, I'm wanting to stream the audio (DJs) from Twitch streams out via [Snapcast](https://github.com/badaix/snapcast), which is over a named FIFO. Looking at the docs, I think I can set it to stream through ffmpeg, and then use ffmpeg to write out to the Snapcast named pipe, right?

Assuming that, seems like I can pass ffmpeg options to have it strip off the video, and only pass the audio along to Snapcast.

Looking ahead at the API, I'd like to write my own [Mopidy](https://mopidy.com/) extension, using Twitch API to see what my current follows are streaming, select one, and start/stop. I imagine that would end up using the Stream object, however it'd still under the hood use VLC/ffmpeg/etc under the hood, behind the scenes, to perform the streaming? Or would I grab/process the data right from Python and send that to the Mopidy output?

**Comment by bastimeyer** [comment, marked as answer]
Twitch already provides audio-only streams for each channel if you select the right stream quality. No remuxing required via ffmpeg. Whether you're using Streamlink's Python API or Streamlink's CLI depends on your use case. You'll have to figure this out on your own.

```
$ streamlink twitch.tv/esl_dota2
[cli][info] Found matching plugin twitch for URL twitch.tv/esl_dota2
Available streams: audio_only, 160p (worst), 360p, 480p, 720p, 720p60, 1080p60 (best)
```

```
$ streamlink --stdout --loglevel=none twitch.tv/esl_dota2 audio_only 2>/dev/null \
  | ffprobe -v error -of json -show_format -show_streams pipe:0 \
  | jq
```
```json
{
  "streams": [
    {
      "index": 0,
      "codec_name": "aac",
      "codec_long_name": "AAC (Advanced Audio Coding)",
      "profile": "LC",
      "codec_type": "audio",
      "codec_tag_string": "[15][0][0][0]",
      "codec_tag": "0x000f",
      "sample_fmt": "fltp",
      "sample_rate": "48000",
      "channels": 2,
      "channel_layout": "stereo",
      "bits_per_sample": 0,
      "initial_padding": 0,
      "id": "0x100",
      "r_frame_rate": "0/0",
      "avg_frame_rate": "0/0",
      "time_base": "1/90000",
      "start_pts": 2796114180,
      "start_time": "31067.935333",
      "bit_rate": "194564",
      "disposition": {
        "default": 0,
        "dub": 0,
        "original": 0,
        "comment": 0,
        "lyrics": 0,
        "karaoke": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "visual_impaired": 0,
        "clean_effects": 0,
        "attached_pic": 0,
        "timed_thumbnails": 0,
        "captions": 0,
        "descriptions": 0,
        "metadata": 0,
        "dependent": 0,
        "still_image": 0
      }
    },
    {
      "index": 1,
      "codec_name": "timed_id3",
      "codec_long_name": "timed ID3 metadata",
      "codec_type": "data",
      "codec_tag_string": "ID3 ",
      "codec_tag": "0x20334449",
      "id": "0x102",
      "r_frame_rate": "0/0",
      "avg_frame_rate": "0/0",
      "time_base": "1/90000",
      "start_pts": 2796114060,
      "start_time": "31067.934000",
      "disposition": {
        "default": 0,
        "dub": 0,
        "original": 0,
        "comment": 0,
        "lyrics": 0,
        "karaoke": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "visual_impaired": 0,
        "clean_effects": 0,
        "attached_pic": 0,
        "timed_thumbnails": 0,
        "captions": 0,
        "descriptions": 0,
        "metadata": 0,
        "dependent": 0,
        "still_image": 0
      }
    }
  ],
  "format": {
    "filename": "pipe:0",
    "nb_streams": 2,
    "nb_programs": 1,
    "format_name": "mpegts",
    "format_long_name": "MPEG-TS (MPEG-2 Transport Stream)",
    "start_time": "31067.934000",
    "bit_rate": "194564",
    "probe_score": 50
  }
}
```

**Reply by Drizzt321** [reply]
Oh wow, nice! I wasn't able to even install this yet, found it just before work, and now need to be doing work for now, but very exciting. Definitely will make it a ton easier. Great starting point, thank you!

**Comment by Drizzt321** [comment]
Just a followup, got it working to a Snapcast pipe input via

`streamlink --stdout --loglevel=none https://www.twitch.tv/viciouscode audio_only 2>/dev/null | ffmpeg -y -i pipe: -f u16le -acodec pcm_s16le -ac 2 -ar 48000 /tmp/streamlinkfifo`

And the Snapcast setup is in the `[stream]` block
`source = pipe:///tmp/streamlinkfifo?name=streamlink`
